config/config.go

```diff
@@ -223,6 +239,104 @@ func (c *Config) Validate() error {
 	return nil
 }
 
+// If the numerator in the Warp config is 0, use the default value
+func setQuorumNumerator(cfgNumerator uint64) uint64 {
+	if cfgNumerator == 0 {
+		return params.WarpDefaultQuorumNumerator
+	}
+	return cfgNumerator
+}
+
+// Helper to retrieve the Warp Quorum from the chain config.
+// Differentiates between subnet-evm and coreth RPC internally
+func getWarpQuorum(
+	subnetID ids.ID,
+	client ethclient.Client,
+) (WarpQuorum, error) {
+	if subnetID == constants.PrimaryNetworkID {
+		return WarpQuorum{
+			QuorumNumerator:   params.WarpDefaultQuorumNumerator,
+			QuorumDenominator: params.WarpQuorumDenominator,
+		}, nil
+	} else {
+		// Fetch the subnet's chain config
+		chainConfig, err := client.ChainConfig(context.Background())
+		if err != nil {
+			return WarpQuorum{}, fmt.Errorf("failed to fetch chain config for subnet %s: %v", subnetID, err)
+		}
+
+		// First, check if the Warp precompile was enabled at genesis
+		warpConfig, ok := chainConfig.GenesisPrecompiles["warpConfig"].(*warp.Config)
+		if ok {
+			return WarpQuorum{
+				QuorumNumerator:   setQuorumNumerator(warpConfig.QuorumNumerator),
+				QuorumDenominator: params.WarpQuorumDenominator,
+			}, nil
+		}
+
+		// If we didn't find the Warp config in the genesis precompile list, check the upgrade precompiles
+		upgrades := chainConfig.ToWithUpgradesJSON().UpgradeConfig.PrecompileUpgrades
+		for _, precompile := range upgrades {
+			warpConfig, ok := precompile.Config.(*warp.Config)
+			if ok {
+				return WarpQuorum{
+					QuorumNumerator:   setQuorumNumerator(warpConfig.QuorumNumerator),
+					QuorumDenominator: params.WarpQuorumDenominator,
+				}, nil
+			}
+		}
+		return WarpQuorum{}, fmt.Errorf("failed to find warp config for subnet %s", subnetID)
+	}
+}
+
+func (c *Config) initializeWarpQuorum() error {
+	c.warpQuorum = make(map[ids.ID]WarpQuorum)
+
+	// Fetch the Warp quorum values for each source subnet
+	for _, sourceSubnet := range c.SourceSubnets {
+		subnetID, err := ids.FromString(sourceSubnet.SubnetID)
+		if err != nil {
+			return fmt.Errorf("invalid subnetID in configuration. error: %v", err)
+		}
+
+		client, err := ethclient.Dial(sourceSubnet.GetNodeRPCEndpoint())
```

Comment:
We should defer closing the client here (after checking the error) and at the dial below